guardian/riff-raff
# refactor: Remove `docker-testkit-*` dependencies in favour of `docker-compose`

## What does this change?



Currently, [`docker-testkit-scalatest` and `docker-testkit-impl-spotify`](https://github.com/whisklabs/docker-it-scala) is used to orchestrate a Docker container running Postgres in test.

In this change, we remove these dependencies in favour of running [`docker-compose`](https://docs.docker.com/compose/reference/) commands:
  - `docker-compose up -d` during test setup to start containers as described in [`docker-compose.yml`](https://github.com/guardian/riff-raff/blob/13fdb406e8ea032f0a3843682abc7be4e84e5aa8/docker-compose.yml) (Postgres)
  - `docker-compose rm --stop --force` during test cleanup to stop and remove the aforementioned containers

[Other projects](https://github.com/search?q=org%3Aguardian+DockerKit&type=code) use `docker-testkit-*` in test. Personally, I'm not a fan of using it in _test_ **and** [`docker-compose` in _local development_](https://github.com/guardian/riff-raff/blob/13fdb406e8ea032f0a3843682abc7be4e84e5aa8/script/start#L12) as:
  - We have to manually keep the content of `docker-compose.yml` and `DockerPostgresService.scala` in sync
  - `docker-testkit-*` does not work on Apple M1 machines (see example `java.lang.UnsatisfiedLinkError` stack trace below)
  - It's an extra set of dependencies
  - It creates a coupling between the tests and Docker

As demonstrated in the commits, I did attempt to bump the version of these libraries, without success.

<details>
<summary> Example java.lang.UnsatisfiedLinkError stack trace</summary>

```
[info] postgres.PostgresDatastoreTest *** ABORTED ***
[info]   com.spotify.docker.client.exceptions.DockerException: java.util.concurrent.ExecutionException: javax.ws.rs.ProcessingException: java.lang.UnsatisfiedLinkError: could not load FFI provider jnr.ffi.provider.jffi.Provider
[info]   at com.spotify.docker.client.DefaultDockerClient.propagate(DefaultDockerClient.java:2848)
[info]   at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:2712)
[info]   at com.spotify.docker.client.DefaultDockerClient.inspectImage(DefaultDockerClient.java:1509)
[info]   at com.whisk.docker.testkit.DockerContainerManager.$anonfun$ensureImage$2(DockerContainerManager.scala:56)
[info]   at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:62)
[info]   at scala.concurrent.package$.blocking(package.scala:124)
[info]   at com.whisk.docker.testkit.DockerContainerManager.$anonfun$ensureImage$1(DockerContainerManager.scala:56)
[info]   at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:678)
[info]   at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
[info]   at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
[info]   ...
[info]   Cause: java.util.concurrent.ExecutionException: javax.ws.rs.ProcessingException: java.lang.UnsatisfiedLinkError: could not load FFI provider jnr.ffi.provider.jffi.Provider
[info]   at jersey.repackaged.com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:299)
[info]   at jersey.repackaged.com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:286)
[info]   at jersey.repackaged.com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:116)
[info]   at com.spotify.docker.client.DefaultDockerClient.request(DefaultDockerClient.java:2710)
[info]   at com.spotify.docker.client.DefaultDockerClient.inspectImage(DefaultDockerClient.java:1509)
[info]   at com.whisk.docker.testkit.DockerContainerManager.$anonfun$ensureImage$2(DockerContainerManager.scala:56)
[info]   at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:62)
[info]   at scala.concurrent.package$.blocking(package.scala:124)
[info]   at com.whisk.docker.testkit.DockerContainerManager.$anonfun$ensureImage$1(DockerContainerManager.scala:56)
[info]   at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:678)
```
</details>

## How to test



- Run the tests locally. They should pass.
- (optional) Run the tests locally on an M1 machine. They should pass.
- Observe CI. It should pass.

## How can we measure success?



- Removal of two dependencies
- Ability to run tests locally on an M1 machine

## Have we considered potential risks?



### Local development
The `docker-compose rm` step happens when running tests locally too. If you've written to the database locally, then run the tests, the database will be removed, and you'll have to start your work again.

I don't think this is a huge problem.

A few possible solutions come to mind:
  - Orchestrate Docker within the [ci script](https://github.com/guardian/riff-raff/blob/main/script/ci), performing cleanup in a [trap](https://www.linuxjournal.com/content/bash-trap-command) to ensure the containers are always removed. This would completely decouple the tests/sbt from Docker.
  - Check for some env vars to determine which environment we're in, and only run the cleanup when in CI. For example if [`teamcity.version`](https://www.jetbrains.com/help/teamcity/predefined-build-parameters.html#Predefined+Server+Build+Parameters) is defined then the tests are running in TeamCity.
  - Have multiple `docker-compose.yml` files (one for the test suite, and one for local development). However that suffers from the problem listed above of having to manually sync changes between them.

### Flaky CI builds?
This might create flaky CI builds. For example:
  - Build A starts, launching Docker containers from cold
  - Build B starts, no need to launch containers as they're already running from above
  - Build A stops, destroying running containers
  - Build B's tests start to fail because the database has disappeared

I don't know enough to say if this is actually a possible scenario. The [docs](https://docs.docker.com/compose/reference/#use--p-to-specify-a-project-name) suggest it might not be as each instance has a project name, defaulted to the directory name. On [TeamCity](https://www.jetbrains.com/help/teamcity/predefined-build-parameters.html#Predefined+Agent+Build+Parameters) this is randomly generated, so builds should already be isolated. If this isn't true, removing the explicit [`container_name`](https://docs.docker.com/compose/compose-file/compose-file-v3/#container_name) might help.
